canonical-web-and-design/canonicalwebteam.discourse-docs
# Restructure

This is a restructure of the codebase to make it easier to do the
pretty URLs work:

- Simplify `DiscourseAPI` so that it is literally just consuming the API
- Separate out the various parsing functions into `parsers.py`
- Surface the meat of the work into the `DiscourseDocs` application itself (it's clearer to see what's going on)
- Rewrite the tests to work with this new structure - testing it at the Flask application level, with HTTPretty

This changes the interface to this module a bit. Now most of the arguments should be provided on the `DiscourseDocs` object, rather than `DiscourseAPI`:

```
DiscourseDocs(
    api=DiscourseAPI(base_url=DISCOURSE_BASE_URL),
    index_topic_id=DOCS_INDEX_TOPIC,
    category_id=DOCS_CATEGORY_ID,
    url_prefix=DOCS_URL_PREFIX,  # Optional
    document_template=DOCS_TEMPLATE_PATH,  # Optional
).init_app(app=app)
```

Nothing much has changed in functionality, but there are two changes:

- Go back to using the `/t/{slug}/{id}` URL path format (this will usually be overridden by the pretty URLs in the coming PR)
- Make the homepage load at `/` rather than `/snap-documentation/3781` (the old URL will redirect to the new)

QA
--

This can be QAed using the [`qa-restructure` branch of docs.snapcraft.io](https://github.com/nottrobin/docs.snapcraft.io/tree/qa-restructure) (whose `requirements.txt` points to this branch), as follows:

``` bash
git clone -b qa-restructure https://github.com/nottrobin/docs.snapcraft.io.git docs.snapcraft.io-qa-restructure
cd docs.snapcraft.io-qa-restructure
./run
```

Now go to http://0.0.0.0:8030, and check:

- http://0.0.0.0:8030/snap-documentation/3781, http://0.0.0.0:8030/t/snap-documentation/3781, http://0.0.0.0:8030/3781, http://0.0.0.0:8030/t/3781 all redirect to http://0.0.0.0:8030/
- http://0.0.0.0:8030/channels/551, http://0.0.0.0:8030t/551, http://0.0.0.0:8030/551 all redirect to http://0.0.0.0:8030/t/channels/551
- Generally, the site works as it used to - including notifications.

Also, check the tests pass on this branch with `./setup.py test`.
